vtex/faststore
# feat(store-ui): Create Carousel component and useSlider hook

## What's the purpose of this pull request?


This PR adds the **Carousel** component and a hook `useSlider` to our `store-ui` library.

## How it works? 


The `Carousel` component should enable users to create carousels of different types of items, such as images (usual image carousel), with the limitation of only showing one item at a time.

![carousel-example](https://user-images.githubusercontent.com/27777263/128728147-afad3eee-758b-4078-b405-2ee567856a7a.png)

The component has the following Props interface:

```ts
interface CarouselProps extends SwipeableProps {
  testId?: string
  infiniteMode?: boolean
  controls?: 'complete' | 'navigationArrows' | 'paginationBullets'
  transition?: {
    duration: number
    property: string
    delay?: number
    timing?: string
  }
}
```

Where `SwipeableProps` is a type from [`react-swipeable`](https://github.com/FormidableLabs/react-swipeable) since we're using this library under the hood and [all options supported by the `useSwipeable` hook](https://github.com/FormidableLabs/react-swipeable/blob/main/src/types.ts#L51) can also be used here.

To enable different CSS rules to be applied to the currently visible items in a given carousel page, we render them with a `data-visible` attribute.

And to offer a convenient way for our users to create their own carousels, product shelves, or pretty much any slider they want without being blocked by the restriction of only being able to display one item at a time, this PR also adds a `useSlider` hook to `store-ui`. The Carouse component uses this hook, and future components that we'll develop, such as the Shelf, will also use it.

## How to test it?


You can check the deploy preview linked below, where you'll find a `Carousel` section listed under `Molecules` in our Storybook.

Since the carousel items are only rendered the first time they're visible, you should only see one image, but then new images should start rendering once you start changing pages.
